Fix `src/test/run-make/issue-36710` on cross-compiled targets

emilyalbini opened the pull request:
This PR fixes the `src/test/run-make/issue-36710` test not working on cross-compiled targets by telling the make infra how to run tests remotely with `remote-test-server`.

This PR includes a revert of #102723 (cc @pcc), which disabled that test on all cross-compiled targets.

emilyalbini committed:
add support for running binaries with remote-test-client on run-make

emilyalbini committed:
Revert "test: run-make: skip when cross-compiling"

This reverts commit 0567fec8e47f83ddda623f93deccddddd3f6744f.

emilyalbini committed:
issue-36710 test can now run on cross-compiled targets

jyn514 commented on src/test/run-make/issue-36710/Makefile:
```diff
@@ -1,6 +1,10 @@
 # ignore-none no-std is not supported
+# ignore-wasm32 FIXME: don't attempt to compile C++ to WASM
+# ignore-wasm64 FIXME: don't attempt to compile C++ to WASM
+# ignore-nvptx64-nvidia-cuda FIXME: can't find crate for `std`
```
Why is this a FIXME? How would CUDA ever get support for libstd? (Is there some way we can say "ignore all targets without std"?)

jyn514 replied:
The commit message adding this originally says https://github.com/rust-lang/rust/commit/1fa48cf181f1fe7e1aba133be199804c652fe55c
> nvtptx64-nvidia-cuda fails in rustc saying it can't find std. The rust
platforms support page says that std is supported on cuda so this is
surprising.

sounds like *something* is broken here but no need to fix it in your PR. cc @RDambrosio016 in case you know who maintains this target.

jyn514 commented on src/test/run-make/issue-36710/Makefile:
```diff
@@ -1,6 +1,10 @@
-# ignore-cross-compile $(call RUN,foo) expects to run the target executable natively
+# ignore-riscv64 $(call RUN,foo) expects to run the target executable natively
```
Why is this still ignored on RISCV?

pietroalbini replied:
This is just a revert of a commit replacing the individual ignores with `ignore-cross-compile`, I did it in a separate step to make it clear what I was changing.

jyn514 commented on src/test/run-make/issue-36710/Makefile:
```diff
@@ -1,6 +1,10 @@
-# ignore-cross-compile $(call RUN,foo) expects to run the target executable natively
+# ignore-riscv64 $(call RUN,foo) expects to run the target executable natively
 #                              so it won't work with remote-test-server
```
What does this mean?

jyn514 reviewed (commented):
Thanks! The code to support cross compiling seems reasonable, but the number of new ignores makes me worried you've missed a target and this will fail on the full test suite. I guess we can just try it and see though.

jyn514 commented:
@bors rollup=iffy

jyn514 commented:
@bors r+

bors commented:
:pushpin: Commit ca362edbdde20d7592643142ea2f00e9f65e2a9e has been approved by `jyn514`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jyn514 ca362edbdde20d7592643142ea2f00e9f65e2a9e -->
<!-- homu: {"type":"Approved","sha":"ca362edbdde20d7592643142ea2f00e9f65e2a9e","approver":"jyn514","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

pietroalbini commented:
Bors seemed to have lost the r+.

@bors r=jyn514

bors commented:
:pushpin: Commit ca362edbdde20d7592643142ea2f00e9f65e2a9e has been approved by `jyn514`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jyn514 ca362edbdde20d7592643142ea2f00e9f65e2a9e -->
<!-- homu: {"type":"Approved","sha":"ca362edbdde20d7592643142ea2f00e9f65e2a9e","approver":"jyn514","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit ca362edbdde20d7592643142ea2f00e9f65e2a9e with merge eab05ae07aab4737b537dd0da052a7bcb2f21e99...
<!-- homu: {"type":"BuildStarted","head_sha":"ca362edbdde20d7592643142ea2f00e9f65e2a9e","merge_sha":"eab05ae07aab4737b537dd0da052a7bcb2f21e99"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3311872386/jobs/5468361149)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3311872386/jobs/5468361149","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`armhf-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3311872386/jobs/5467850185) [(plain)](https://github.com/rust-lang-ci/rust/commit/eab05ae07aab4737b537dd0da052a7bcb2f21e99/checks/9066366875/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [run-make] src/test/run-make/rustdoc-scrape-examples-ordering ... ok

failures:

---- [run-make] src/test/run-make/issue-36710 stdout ----
error: make failed
status: exit status: 2
command: "make"
--- stdout -------------------------------
--- stdout -------------------------------
arm-linux-gnueabihf-g++ -ffunction-sections -fdata-sections -fPIC -march=armv6 -marm -mfpu=vfp -c -o /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/issue-36710/issue-36710/libfoo.o foo.cpp
--- stderr -------------------------------
--- stderr -------------------------------
/bin/dash: 1: arm-linux-gnueabihf-g++: not found
make: *** [Makefile:17: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/issue-36710/issue-36710/libfoo.o] Error 127



failures:
```

</details>

emilyalbini committed:
install the target g++ in armhf-gnu

pietroalbini commented:
Installed the missing compiler in that builder, hopefully this is the only builder where it's missing :crossed_fingers:

jyn514 commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit cb096c148285f5f1146ec1a61cf0b297c78ec473 has been approved by `jyn514`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jyn514 cb096c148285f5f1146ec1a61cf0b297c78ec473 -->
<!-- homu: {"type":"Approved","sha":"cb096c148285f5f1146ec1a61cf0b297c78ec473","approver":"jyn514","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit cb096c148285f5f1146ec1a61cf0b297c78ec473 with merge fa8492c23ff5f8d57adf27c88dda77a1477623c8...
<!-- homu: {"type":"BuildStarted","head_sha":"cb096c148285f5f1146ec1a61cf0b297c78ec473","merge_sha":"fa8492c23ff5f8d57adf27c88dda77a1477623c8"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3355662334/jobs/5560332683)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3355662334/jobs/5560332683","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
The job **`arm-android`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3355662334/jobs/5560097729) [(plain)](https://github.com/rust-lang-ci/rust/commit/fa8492c23ff5f8d57adf27c88dda77a1477623c8/checks/9187843673/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [run-make] src/test/run-make/rustdoc-scrape-examples-ordering ... ok

failures:

---- [run-make] src/test/run-make/issue-36710 stdout ----
error: make failed
status: exit status: 2
command: "make"
--- stdout -------------------------------
--- stdout -------------------------------
arm-linux-androideabi-clang++ -DANDROID -ffunction-sections -fdata-sections -fPIC -c -o /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/issue-36710/issue-36710/libfoo.o foo.cpp
--- stderr -------------------------------
--- stderr -------------------------------
/bin/dash: 1: arm-linux-androideabi-clang++: not found
make: *** [Makefile:17: /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/issue-36710/issue-36710/libfoo.o] Error 127



failures:
```

</details>

emilyalbini committed:
set correct default value for cc and cxx on android

emilyalbini committed:
run tests on the remote device even when the default address is used

When running tests inside the Android emulator, bootstrap doesn't set
the TEST_DEVICE_ADDR environment variable, as the default address
(127.0.0.1:12345) is used.

Instead, REMOTE_TEST_CLIENT is set all the times when remote testing is
needed, and in no other cases. To ensure Android tests are executed in
the emulator, change the check.

pietroalbini commented:
Android failed because of two problems:

* When configuring the NDK we were not setting the `cc` and `cxx` bootstrap configs to point to the NDK's compilers. This didn't affect building code with Cargo, but resulted in the `CC` and `CXX` environment variables not being set correctly in `run-make` tests.
* Android used the default value for `TEST_DEVICE_ADDR` and so didn't set the environment variable, breaking the detection logic I had in `run-make`.

I fixed both of the problems and successfully ran the `run-make` suite locally with `src/ci/docker/run.sh`.

bors commented:
:umbrella: The latest upstream changes (presumably #104188) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jyn514 commented:
Sorry for the delay. r=me after a rebase.

bors commented:
:pushpin: Commit 6bfbd113e25f83761a232404e6ff71af5d07614a has been approved by `jyn514`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jyn514 6bfbd113e25f83761a232404e6ff71af5d07614a -->
<!-- homu: {"type":"Approved","sha":"6bfbd113e25f83761a232404e6ff71af5d07614a","approver":"jyn514","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/104361#issuecomment-1317416027) for pull requests below priority 1. This pull request will be tested once the tree is reopened.

emilyalbini committed:
fmt

pietroalbini commented:
@bors r=jyn514

bors commented:
:pushpin: Commit 00ec6797a3b9f5e0fe3c6e66af0a3a6c82fa56c0 has been approved by `jyn514`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jyn514 00ec6797a3b9f5e0fe3c6e66af0a3a6c82fa56c0 -->
<!-- homu: {"type":"Approved","sha":"00ec6797a3b9f5e0fe3c6e66af0a3a6c82fa56c0","approver":"jyn514","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:evergreen_tree: The tree is currently [closed](https://github.com/rust-lang/rust/pull/104361#issuecomment-1317416027) for pull requests below priority 1. This pull request will be tested once the tree is reopened.

bors commented:
:hourglass: Testing commit 00ec6797a3b9f5e0fe3c6e66af0a3a6c82fa56c0 with merge 30117a1dbb843da1d5ab1258e89a3ed0b1940475...
<!-- homu: {"type":"BuildStarted","head_sha":"00ec6797a3b9f5e0fe3c6e66af0a3a6c82fa56c0","merge_sha":"30117a1dbb843da1d5ab1258e89a3ed0b1940475"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3494833575/jobs/5853349984)
Approved by: jyn514
Pushing 30117a1dbb843da1d5ab1258e89a3ed0b1940475 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"jyn514","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3494833575/jobs/5853349984"},"merge_sha":"30117a1dbb843da1d5ab1258e89a3ed0b1940475"} -->

rust-timer commented:
Finished benchmarking commit (30117a1dbb843da1d5ab1258e89a3ed0b1940475): [comparison URL](https://perf.rust-lang.org/compare.html?start=83356b78c4ff3e7d84e977aa6143793545967301&end=30117a1dbb843da1d5ab1258e89a3ed0b1940475&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=83356b78c4ff3e7d84e977aa6143793545967301&end=30117a1dbb843da1d5ab1258e89a3ed0b1940475&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=83356b78c4ff3e7d84e977aa6143793545967301&end=30117a1dbb843da1d5ab1258e89a3ed0b1940475&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 3.7% | [2.5%, 5.0%] | 2     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=83356b78c4ff3e7d84e977aa6143793545967301&end=30117a1dbb843da1d5ab1258e89a3ed0b1940475&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.
